portfolio/src/main/java/com/google/sps/servlets/FormHandlerServlet.java

```diff
@@ -0,0 +1,289 @@
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Message> messages = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String firstName = (String) entity.getProperty("firstName");
+      String lastName = (String) entity.getProperty("lastName");
+      String text = (String) entity.getProperty("text");
+      Double sentimentScore = (Double) entity.getProperty("sentimentScore");
+      String sentimentScoreStr = String.format("%.2f", sentimentScore);
+      String imageUrl = (String) entity.getProperty("imageUrl");
+      List<String> imageLabelDescriptions = 
+          new ArrayList<>((Collection<String>) entity.getProperty("imageLabelDescriptions"));
+      List<String> imageLabelScores = 
+          new ArrayList<>((Collection<String>) entity.getProperty("imageLabelScores"));
+
+      String emojiCode;
+      if (sentimentScore < -0.5) {
+        // angry face
+        emojiCode = "0x1F620";
+      } else if (sentimentScore < 0) {
+        // slightly frowning face
+        emojiCode = "0x1F641";
+      } else if (sentimentScore == 0) {
+        // neutral face
+        emojiCode = "0x1F610";
+      } else if (sentimentScore < 0.5) {
+        // slightly smiling face
+        emojiCode = "0x1F642";
+      } else {
+        // grinning face with big eyes
+        emojiCode = "0x1F603";
+      }
+
+      String combinedText = text + " - " + firstName + " " + lastName +
+          " (Sentiment Score: " + sentimentScoreStr + ")";
+      
+      Message message = new Message(combinedText, emojiCode, imageUrl, imageLabelDescriptions,
+          imageLabelScores);
+      messages.add(message);
+    }
+
+    // Convert the messages to JSON.
+    Gson gson = new Gson();
+    String json = gson.toJson(messages);
+
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
+    PrintWriter out = response.getWriter();
+
+    // Get the input from the form.
+    String firstName = getParameter(request, "f-name", "");
+    String lastName = getParameter(request, "l-name", "");
+    String text = getParameter(request, "text-input", "");
+
+    // Get the BlobKey that points to the image uploaded by the user.
+    BlobKey blobKey = getBlobKey(request, "image");
+
+    // User didn't upload a file, so render an error message.
+    if (blobKey == null) {
+      out.println("Please upload an image file.");
+      return;
+    }
+
+    // Get the URL of the image that the user uploaded to Blobstore.
+    String imageUrl = getUploadedFileUrl(blobKey);
+
+    // Get the labels of the image that the user uploaded.
+    byte[] blobBytes = getBlobBytes(blobKey);
+    List<EntityAnnotation> imageLabels = getImageLabels(blobBytes);
+
+    // Create two separate lists for label descriptions and label scores
```

Comment:
When I see two lists that hold related data, usually that's a sign that I should create a new class and only have one list.

Consider creating something like an `ImageLabel` class that contains a label and its score, and then only using one list.

Reply:
:O good idea!

Reply:
It looks like the new class I create is not a supported property type for an entity in the Datastore. I remember I got the same error when I tried to do 
setProperty("imageLabels", imageLabels) for the list of EntityAnnotations, which was why I split it into two lists of Strings.

Reply:
ok I decided to store the BlobKey instead, and that worked